File: IcybeckaImageUtil/unused/reducecolors.c

#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include <raylib.h>

#define min(a,b) ((a)<(b):(a):(b))
#define max(a,b) ((a)>(b):(a):(b))

uint32_t reduceColors(uint32_t color, uint8_t removebits, uint8_t cmin, uint8_t cmax) {
	uint8_t r = (color >>  0) & 0xff;
	uint8_t g = (color >>  8) & 0xff;
	uint8_t b = (color >> 16) & 0xff;
	uint8_t a = (color >> 24) & 0xff;
	if (r < cmin) r = 0;
	if (g < cmin) g = 0;
	if (b < cmin) b = 0;
	if (a < cmin) a = 0;
	if (r > cmax) r = 0xff;
	if (g > cmax) g = 0xff;
	if (b > cmax) b = 0xff;
	if (a > cmax) a = 0xff;
	if (r > 0 && r < 0xff) r &= 0xff << removebits;
	if (g > 0 && g < 0xff) g &= 0xff << removebits;
	if (b > 0 && b < 0xff) b &= 0xff << removebits;
	if (a > 0 && a < 0xff) a &= 0xff << removebits;
	return r | (g << 8) | (b << 16) | (a << 24);
}

Image reduceImageColorBits(Image img, uint8_t removebits, uint8_t cmin, uint8_t cmax) {
	size_t numcolors;
	size_t *counts;
	ImageFormat(&img, PIXELFORMAT_UNCOMPRESSED_R8G8B8A8);
	for (size_t y = 0; y < img.height; y++) {
		for (size_t x = 0; x < img.width; x++) {
			((uint32_t*)img.data)[x + y*img.width] = reduceColors(((uint32_t*)img.data)[x + y*img.width], removebits, cmin, cmax);
		}
	}
	return img;
}

int readint(const char *s) {
	char c;
	size_t i = 0;
	int num = 0;
	do {
		c = s[i++];
		if (c >= '0' && c <= '9') {
			num = num * 10 + c - '0';
		}
	} while (c);
	return num;
}

int main(int argc, char **argv) {
	int removebits, cmin, cmax;
	Image img;
	if (argc < 6) {
		printf("Usage: %s image_in image_out reduce_bits channel_min channel_max\nReduces colors from image_in to image_out. channel_min is the value per channel that is treated as minimum, channel_max is the value per channel that is treated as maximum. reduce_bits is the number of bits to chop off the value otherwise.\nreduce_bits: 0-7\nchannel_min: 0-255\nchannel_max: 0-255\n", argv[0]);
		return 0;
	}
	img = LoadImage(argv[1]);
	if (!IsImageReady(img)) {
		printf("Failed to open image file %s\n", argv[1]);
		return -1;
	}
	removebits = readint(argv[3]);
	cmin = readint(argv[4]);
	cmax = readint(argv[5]);
	if (removebits < 0 || removebits > 7) {
		printf("reduce_bits out of range. (0-7)\n");
		return 1;
	}
	if (cmin < 0 || cmin > 0xff) {
		printf("cmin out of range. (0-255)\n");
		return 1;
	}
	if (cmax < 0 || cmax > 0xff) {
		printf("cmax out of range. (0-255)\n");
		return 1;
	}
	img = reduceImageColorBits(img, removebits, cmin, cmax);
	if (!ExportImage(img, argv[2])) {
		printf("Failed to write output image file %s\n", argv[2]);
		return -2;
	}
	return 0;
}

File: IcybeckaImageUtil/reducecolors.py

from PIL import Image
import os

def walk(d, types=None):
	for root,dirs,files in os.walk(d):
		for f in files:
			if types is None or f.rsplit('.', maxsplit=1)[-1].lower() in types:
				yield os.path.join(root, f)

def reduceImageColors(img, removebits=0, cmin=0, cmax=0xff):
	px = img.load()
	for y in range(img.height):
		for x in range(img.width):
			o = []
			for c in list(px[x, y]):
				if c > 0 and c < cmin: c = cmin
				if c < 255 and c > cmax: c = cmax
				if c > 0 and c < 255: c &= 0xff * 2 ** removebits
				o.append(c)
			px[x, y] = tuple(o)
	return img

if __name__=='__main__':
	import sys
	if len(sys.argv) < 6:
		print(f"Usage: {sys.argv[0]} image_in image_out reduce_bits channel_min channel_max [output_format [input_formats]]\nReduces colors from image_in to image_out.\nIf image_in is a directory, all images within will be reduced into directory specified by image_out.\nchannel_min is the value per channel that is treated as minimum, channel_max is the value per channel that is treated as maximum.\nreduce_bits is the number of bits to chop off the value otherwise.\nreduce_bits: 0-7\nchannel_min: 0-255\nchannel_max: 0-255\noutput_format is the format (such as png) to output.\ninput_formats is a list of comma-delimited file formats to input from when image_in is a directory. example: png,jpg,bmp\n")
		exit(0)

	if len(sys.argv) > 6:
		output_format = sys.argv[6]
		if not output_format.startswith("."):
			output_format = "." + output_format
	else:
		output_format = None

	if len(sys.argv) > 7:
		image_formats = sys.argv[7].split(",")
	else:
		image_formats = ["blp","bmp","dds","eps","gif","icns","ico","im","jpeg","jpg",
						"j2k","j2p","jpx","msp","pcx","png","apng","ppm","sgi","spider",
						"tga","tiff","webp","xbm","cur","dcx","fits","fli","flc","fpx",
						"ftex","gbr","imt","iptc","naa","mcidas","mic","mpo","pcd","pixar",
						"psd","qoi","sun","wal","wmf","emf","xpm"]

	removebits = int(sys.argv[3])
	cmin = int(sys.argv[4])
	cmax = int(sys.argv[5])
	if removebits < 0 or removebits > 7:
		print("remove_bits out of range. (0-7)")
		exit(1)
	if cmin < 0 or cmin > 255:
		print("cmin out of range. (0-255)")
		exit(1)
	if cmax < 0 or cmax > 255:
		print("cmax out of range. (0-255)")
		exit(1)
	
	fnames = [(sys.argv[1], sys.argv[2])]
	if os.path.isdir(sys.argv[1]):
		try:
			os.makedirs(sys.argv[2])
		except IOError:
			pass
		except Exception as e:
			print(f"Failed to create directory {sys.argv[2]}\nOriginal error: {e}")
			exit(1)
		if output_format is None:
			fnames = [(f, os.path.join(sys.argv[2], os.path.basename(f))) for f in walk(sys.argv[1], image_formats)]
		else:
			fnames = [(f, os.path.join(sys.argv[2], os.path.basename(f).rsplit(".", maxsplit=1)[0]+output_format)) for f in walk(sys.argv[1], image_formats)]

	for fname, dname in fnames:
		try:
			print(f"Reducing {fname} -> {dname}")
			img = Image.open(fname)
			img = reduceImageColors(img, removebits, cmin, cmax)
			if removebits >= 4 and "A" in img.channels:
				img = img.convert("")
			img.save(dname, optimize=True)
		except FileNotFoundError:
			print(f"File {fname} not found")
		except Exception as e:
			print(f"Failed to reduce image file {fname}\nOriginal error: {e}")

